src/app/core/services/kafka.service.ts

```diff
@@ -35,82 +42,85 @@ export class KafkaService {
-      )
-    })
+    return this.prepareKafkaObject(task, Answer, KAFKA_ASSESSMENT)
   }
 
   prepareNonReportedTasksKafkaObject(task: Task) {
     // NOTE: Payload for kafka 1 : value Object which contains individual questionnaire response with timestamps
     const CompletionLog: CompletionLogValueExport = {
       name: task.name.toString(),
-      time: task.timestamp,
+      time: task.timestamp / SEC_MILLISEC,
       completionPercentage: { double: task.completed ? 100 : 0 }
     }
+    return this.prepareKafkaObject(task, CompletionLog, KAFKA_COMPLETION_LOG)
+  }
+
+  prepareTimeZoneKafkaObject() {
+    const ApplicationTimeZone: ApplicationTimeZoneValueExport = {
+      time: new Date().getTime() / SEC_MILLISEC,
+      offset: new Date().getTimezoneOffset() * MIN_SEC
+    }
+    return this.prepareKafkaObject([], ApplicationTimeZone, KAFKA_TIMEZONE)
+  }
+
+  getSpecs(task: Task, kafkaObject, type) {
+    switch (type) {
+      case KAFKA_ASSESSMENT:
+        return this.storage.getAssessmentAvsc(task).then(specs => {
+          return Promise.resolve(
+            Object.assign(specs, { task: task, kafkaObject: kafkaObject })
+          )
+        })
+      case KAFKA_COMPLETION_LOG:
+      case KAFKA_TIMEZONE:
+        return Promise.resolve({
+          name: type,
+          avsc: 'questionnaire',
+          task: task,
+          kafkaObject: kafkaObject
+        })
+      default:
+        break
+    }
+  }
 
-    this.util.getSourceKeyInfo().then(keyInfo => {
+  prepareKafkaObject(task, value, type) {
+    return this.util.getSourceKeyInfo().then(keyInfo => {
       const sourceId = keyInfo[0]
       const projectId = keyInfo[1]
       const patientId = keyInfo[2].toString()
-      const AnswerKey: AnswerKeyExport = {
+      // NOTE: Payload for kafka 2 : key Object which contains device information
+      const answerKey: AnswerKeyExport = {
         userId: patientId,
         sourceId: sourceId,
         projectId: projectId
       }
-      const kafkaObject = { value: CompletionLog, key: AnswerKey }
-      this.getSpecs(task, kafkaObject).then(specs =>
-        this.createPayload(specs, task, kafkaObject)
+      const kafkaObject = { value: value, key: answerKey }
+      return this.getSpecs(task, kafkaObject, type).then(specs =>
+        this.createPayload(specs, task, kafkaObject, type)
       )
     })
   }
 
-  getSpecs(task: Task, kafkaObject) {
-    if (kafkaObject.value.completionPercentage !== undefined) {
-      return Promise.resolve({
-        name: 'completion_log',
-        avsc: 'questionnaire',
-        task: task,
-        kafkaObject: kafkaObject
-      })
-    } else {
-      return this.storage.getAssessmentAvsc(task).then(specs => {
-        return Promise.resolve(
-          Object.assign(specs, { task: task, kafkaObject: kafkaObject })
-        )
-      })
-    }
-  }
-
-  createPayload(specs, task, kafkaObject) {
+  createPayload(specs, task, kafkaObject, type) {
```

Comment:
Perhaps a better name like `createPayloadAndSend()` ? Or maybe break the login into 2 functions, one for payload an done for sending. 
Same naming for `prepareAnswerKafkaObject()` like `prepareAnswerKafkaObjectAndSend()`?

Reply:
Made the changes.